core/arch/arm/plat-synquacer/rng_pta.c

```diff
@@ -0,0 +1,211 @@
+	0x40df0b66, 0x37d83bf0, 0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
+	0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6, 0xbad03605, 0xcdd70693,
+	0x54de5729, 0x23d967bf, 0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
+	0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
+};
+
+static uint32_t crc32(uint8_t *msg, uint32_t sz)
+{
+	uint32_t i, crc = 0xffffffff;
+
+	for (i = 0; i < sz; i++)
+		crc = (crc >> 8) ^ table[(crc ^ msg[i]) & 0xff];
+
+	return ~crc;
+}
+
+static void pool_add_entropy(uint8_t byte)
+{
+	entropy_pool[entropy_size] = byte;
+	entropy_size++;
+}
+
+static void pool_get_entropy(uint8_t *buf, uint32_t size)
+{
+	uint32_t off;
+
+	if (size > entropy_size)
+		return;
+
+	off = entropy_size - size;
+
+	memcpy(buf, &entropy_pool[off], size);
+	entropy_size -= size;
+}
+
+static uint32_t pool_get_entropy_sz(void)
+{
+	uint32_t sz;
+
+	sz = entropy_size;
+
+	return sz;
+}
+
+void rng_collect_entropy(void)
+{
+	uint64_t ts0 = 0, ts1 = 0, ts2 = 0, ts3 = 0, ts4 = 0, ts5 = 0, ts6 = 0;
+	uint8_t e = 0;
+
+	ts0 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE0 +
+					      TEMP_DATA_REG_OFFSET));
+	ts1 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE1 +
+					      TEMP_DATA_REG_OFFSET));
+	ts2 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE2 +
+					      TEMP_DATA_REG_OFFSET));
+	ts3 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE3 +
+					      TEMP_DATA_REG_OFFSET));
+	ts4 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE4 +
+					      TEMP_DATA_REG_OFFSET));
+	ts5 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE5 +
+					      TEMP_DATA_REG_OFFSET));
+	ts6 = read32((vaddr_t)phys_to_virt_io(THERMAL_SENSOR_BASE6 +
+					      TEMP_DATA_REG_OFFSET));
+
+	e = ((ts0 & 0x1) << 6) | ((ts1 & 0x1) << 5) | ((ts2 & 0x1) << 4) |
+	    ((ts3 & 0x1) << 3) | ((ts4 & 0x1) << 2) | ((ts5 & 0x1) << 1) |
+	    ((ts6 & 0x1) << 0);
+
+	e |= ((ts0 ^ ts1 ^ ts2 ^ ts3 ^ ts4 ^ ts5 ^ ts6) << 6) & 0x80;
+
+	pool_add_entropy(e);
+
+	if (pool_get_entropy_sz() >= ENTROPY_POOL_SIZE)
+		generic_timer_stop();
+}
+
+static TEE_Result rng_get_entropy(uint32_t types,
+				  TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint8_t *e = NULL;
+	uint32_t i, digest, pool_size = 0, rq_size = 0;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	rq_size = params[0].memref.size;
+
+	if ((rq_size == 0) || (rq_size > ENTROPY_POOL_SIZE))
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	e = (uint8_t *)params[0].memref.buffer;
+	if (e == NULL)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	pool_size = pool_get_entropy_sz();
+
+	if (pool_size < rq_size) {
+		params[0].memref.size = pool_size;
+		pool_get_entropy(e, pool_size);
+	} else {
+		params[0].memref.size = rq_size;
+		pool_get_entropy(e, rq_size);
+	}
+
+	/* CRC whitening */
```

Comment:
Yes CRC whitening seems to be essential as per our randomness measurement using FIPS RNG fitness tests. Implementing it in secure world seems more efficient than implementing it in various clients like boot-loader, OS etc.
Regarding the last bytes, providing CRC lesser than 4 bytes as input would add bias to output 4 bytes of RNG data. So I have left them intact intentionally.

Reply:
It's still a fairly raw randomness and mixing CRC with security seems like a bad combination. This is a bit of a home brew solution, I'd rather not take the responsibility of that.

As long as we're not using a well established method it's better to be honest in the interface and leave the responsibility to the client to turn the entropy into something that can be used as a random number.

Reply:
+1

Why not just feed the raw data from the thermal sensors into the existing Fortuna PRNG and be done with it? (and of course stop claiming it is a "True RNG"). IMO it would be as good as it gets.

Reply:
If we were restricting ourselves to implementing the GP internal API I think a CSPRNG might be the best option due to the slow bit rate of the this source (the API looked like we would have to wait until we accumulate enough data and that could easily lead to be unacceptably long waits for some applications).

However one of the most obvious use cases for the SMC interface is to provide the Linux kernel with entropy... and the kernel already uses the entropy (by default extremely conservatively) to seed a CSPRNG. Thus for this use case placing another CSPRNG in between isn't very useful.

Likewise conditioning components are sometimes tailored to the idiosyncrasies of the noise source so pushing it out completely raw numbers to the driver means splitting the knowledge of the entropy source design between OP-TEE and kernel which is also not great.

However after (more) reading of some of the NIST standards I have noted:

1. The SMC interface does not allow the entropy source to describe its entropy level (hence it is implicitly claiming 8-bits per byte)
2. There are NIST frameworks by which we could calculate the entropy level of a "non vetted" conditioning component such as crc32() but the framework is such that a non-vetted conditioner can never achieve 8-bits per byte
3. We could switch to a vetted conditioner such as one based on SHA512/256 (or SHA3)

Finally it strikes me also that the SMC interface could provide choice (GetNoise(), GetEntropy(), GetSeededCSPRNG() ). The last of these isn't useful to the kernel (which can look after itself) but could find uses in simpler environments such as bootloaders.

Reply:
@jforissier @jenswi-linaro It seems Fortuna CSPRNG implementation would be useful here to feed raw thermal sensor data directly. So I will remove CRC whitening. Along with that we would like to keep an additional interface to provide un-whitened data to normal world.

Reply:
What's the purpose of the un-whitened data? Any reason why CSPRNG can't be used instead?

Reply:
We just want to have actual entropy source for normal world to experiment with.

Reply:
So it's for testing purposes. I guess it's fine if it's disabled by default and it's clear that it's not for production use.

Reply:
Will create a config option for this.

Reply:
> I think we can disregard the kernel in this design since
> it doesn't matter to the kernel if it receives its seed from
> a CSPRNG or something else.

The kernel *does* distinguish between RDRAND (CSPRNG) versus RDSEED (TRNG) but that's mostly for the architecture defined RNGs.

For non-architectural RNGs everything slots into the hw_random driver family and AFAICT all the of the current drivers are attached to hardware that claims to be a TRNG (e.g. no entropy "stretching").


> What's the purpose of the un-whitened data? Any reason why
> CSPRNG can't be used instead?

Access to un-whitened data is essentially the GetNoise() function from NIST SP 800-90B: *This
interface is meant to provide test data to credit a noise source with an entropy estimate during
validation or for external health testing.*